data-library/library/script/nycoc_checkbook.py

```diff
@@ -0,0 +1,381 @@
+        self,
+        user_agent: str | None = None,
+        type_of_data: TYPE_OF_DATA = "Spending",
+        records_from: str = DEFAULT_RECORDS_FROM,
+        max_records: str = DEFAULT_MAX_RECORDS,
+        response_columns: list[str] | None = None,
+        search_criteria: list[CriteriaValue | CriteriaRange] = [],
+        num_retries: int = 3,
+        data_dir: Path = DEFAULT_DATA_PATH,
+    ) -> pd.DataFrame:
+        """Fetches data from Checkbook NYC API based on input parameters, saves it
+        locally in parquet files organized by months, and returns data in a pandas DataFrame.
+        """
+        create_data_dir(data_dir)
+
+        start_period = DEFAULT_START_PERIOD
+        end_period = DEFAULT_END_PERIOD
+
+        # check if search_criteria has "issue_date" filter. If not, use default values.
+        if search_criteria:
+            for index, criteria_dict in enumerate(search_criteria):
+                if (
+                    criteria_dict["type"] == "range"
+                    and criteria_dict["name"] == "issue_date"
+                ):
+                    start_period = criteria_dict["start"]
+                    end_period = criteria_dict["end"]
+                    search_criteria.pop(index)
+
+        month_range = generate_monthly_ranges(start_period, end_period)
+
+        for month in month_range:
+            start, end = month
+            month_criteria: CriteriaRange = {
+                "name": "issue_date",
+                "type": "range",
+                "start": start,
+                "end": end,
+            }
+            search_criteria_modified = search_criteria + [month_criteria]
+            df = get_response(
+                user_agent=user_agent,
+                type_of_data=type_of_data,
+                records_from=records_from,
+                max_records=max_records,
+                response_columns=response_columns,
+                search_criteria=search_criteria_modified,
+                num_retries=num_retries,
+            )
+            if df is not None:
+                filename = f"{start}-{end}_checkbook.parquet"
+                filepath = str(data_dir / filename)
+                df.to_parquet(filepath)
+            else:
+                print(f"Unable to extract data for period: {start}-{end}...")
+
+            # pausisng before next request: we don't want to overload API
+            time.sleep(random.randint(5, 10))
+
+        data = read_parquet_files_to_df(data_dir)
+        print("Successfully ingested data.")
+        return data
+
+    def runner(self) -> str:
+        df = self.ingest(
+            user_agent=user_agent,
+            search_criteria=[
+                {"name": "spending_category", "type": "value", "value": "cc"}
+            ],
+        )
+        local_path = df_to_tempfile(df)
+        return local_path
+
+
+def get_response(
+    user_agent: str | None,
+    type_of_data: TYPE_OF_DATA,
+    records_from: str,
+    max_records: str,
```

Comment:
@sf-dcp Are there ever cases where there are more than `max_records` (ie 20k) within a given month range? Curious what would happen in that case?

Reply:
@alexrichey, it seems that the average number of records per month is between 10-15k. 

If the count is greater than 20k, the current logic will return the first 20k. A potential enhancement is reading in the  total count of records for a given month (it's returned in the response body) and make additional API requests, using the `records_from` parameter. For example, let's say the total count is 35k; then we can send 2 requests where first request will return records 1-20,000 and second request - 20,001-35,000.

Reply:
I'd say that's an enhancement worth getting in this round. Or at a bare minimum, throwing an error if the record count is greater than 20000, but that's just kicking the issue down the road a bit. Possibility of silently missing data is one that we want to avoid

Reply:
@sf-dcp that sounds like a great approach. Let's hope we don't find a day in which >20k checks were cut... though maybe there are other ways to filter down the data in addition

Reply:
Added enhancements to the code.